Python3 installer rewrite

bleggett opened the pull request:
Ready for review/feedback.

- Rewrites installer wizard and most nontrivial shell scripts in Python3.
- Uses `pip` and `requirements.txt` up front for dependency management.
- Implements a CLI installer using Click and nested commands and subcommands `Installer.py [opts] install [installopts] scenario3 [scenario3opts]
- Click installer will prompt user for input if a required option is missing, but all options can be explicitly supplied via command-line arg, to make noninteractive execution possible.
- Reorganizes the installer folder structure in the following (hopefully more obvious) layout:
  - jazz-installer
    - (platform-specific prerequisite installer scripts)
    - feature-extensions
    - installer
      - cli (the installer logic, Click modules and submodules)
      - configurators (code that needs to configure Jenkins/Gitlab/Sonar/etc before the Terraform execution step)
      - helpers (general shared functions)
      - terraform (all Terraform resources are in this folder, including external scripts that Terraform invokes)

#### Excerpt from updated user guide for context below:
=================================================

## User Guide

Quick version:

Ensure you have a standard install of Python 3.6 or greater, with `pip` and `setuptools`

Run:

```
git clone git@github.com:tmobile/jazz-installer.git
cd jazz-installer
python3 -m virtualenv env && source env/bin/activate
pip install -r requirements.txt
python Installer.py install --stackprefix myteststack --adminemail test@tst.com --region us-east-1 scenario3
```

Uninstall:
```
cd jazz-installer
source env/bin/activate
python Installer.py uninstall --mode=[all|frameworkonly]
```

Options:
```
python Installer.py
```


For more details, see the [Wiki](https://github.com/tmobile/jazz-installer/wiki).

bleggett commented:
This is stuck for the time being as one of our internal AWS instances I was using for testing is on lockdown and I attempt to get set up elsewhere.

bleggett commented:
Currently testing this with scenario3, I will see about scheduling and testing scenario 1/2 runs after that is done, but I'm trying to wrap this up.

bleggett commented:
Excerpt from updated user guide for context below:
=================================================

## User Guide

Quick version:

Ensure you have a standard install of Python 3.6 or greater, with `pip` and `setuptools`

Run:

```
git clone git@github.com:tmobile/jazz-installer.git
cd jazz-installer
(optional if you want a venv: python3 -m virtualenv env && source env/bin/activate)
pip install -r requirements.txt
python Installer.py install scenario3
```

For more details, see the [Wiki](https://github.com/tmobile/jazz-installer/wiki).

bleggett commented:
tested working install/uninstall with scenario3, will test with scenario 1 next

suryajak commented on installer/cli/click_required.py:
```diff
@@ -0,0 +1,45 @@
+import click
+
+
+# https://stackoverflow.com/questions/44247099/click-command-line-interfaces-make-options-required-if-other-optional-option-is
```
This is neat!

suryajak commented on installer/cli/scenario1cli.py:
```diff
@@ -0,0 +1,144 @@
+@click.option(
+    '--bitbucket_endpoint',
+    required=True,
+    help='Provide the endpoint (without the protocol or port) of the Bitbucket server to be configured',
```
"Provide the DNS name ..." so that we are consistent

suryajak commented on installer/cli/scenario1cli.py:
```diff
@@ -0,0 +1,144 @@
+)
+@click.option(
+    '--sonarqube_endpoint',
+    help='Provide the endpoint (without the protocol or port) of the Sonarqube server to be configured',
```
"Provide the DNS name ..." so that we are consistent

suryajak commented on installer/cli/scenario1cli.py:
```diff
@@ -0,0 +1,144 @@
+    help='Provide the username and password of the existing Sonarqube user separated by a space'
+)
+@click.option(
+    '--sonarqube_publicip',
```
I don't remember why this is required when we already have the DNS name. If the existing code has it, then lets keep it.

bleggett replied:
Existing code does have this, I do not know either, but especially for scenario1 I'm just porting stuff over as-is.

raghits replied:
@bleggett Can this be marked resolved?

suryajak commented on installer/cli/scenario2cli.py:
```diff
@@ -0,0 +1,73 @@
+@click.option(
+    '--bitbucket_publicip',
+    required=True,
+    help='Provide the public IP of the Bitbucket server to be configured.',
```
"Provide the DNS name ..." so that we are consistent

bleggett replied:
Fixed.

suryajak commented on installer/cli/uninstallcli.py:
```diff
@@ -0,0 +1,33 @@
+@click.option(
+    '--mode',
+    type=click.Choice(['all', 'frameworkonly']),
+    help='`all` to remove Jazz and all deployed functions, \
```
"all deployed services" as we have API's, Fns and SPAs.

bleggett replied:
Fixed.

suryajak commented on installer/cli/uninstallcli.py:
```diff
@@ -0,0 +1,33 @@
+    '--mode',
+    type=click.Choice(['all', 'frameworkonly']),
+    help='`all` to remove Jazz and all deployed functions, \
+    `frameworkonly` to remove Jazz but leave deployed functions alone',
```
Same comment as above - "all deployed services" as we have API's, Fns and SPAs.

bleggett replied:
Fixed.

suryajak commented on installer/configurators/common.py:
```diff
@@ -0,0 +1,130 @@
+            exit(0)
+
+
+# Uses sed to modify the values of key-value pairs within a file
```
Comment is no longer applicable. Also the old code below can go away too.

bleggett replied:
Yep, thx.

suryajak commented on installer/configurators/sonarqube.py:
```diff
@@ -0,0 +1,47 @@
+        )
+    else:
+        sys.exit(
+            "Kindly provide an 'Admin' Sonar user with correct password and run the installer again!"
```
It sucks that the user has to start over. Could we perhaps ask the user to re-share their admin credentials and run the check_sonar_user?

bleggett replied:
It does, but again, modelling existing code here. We can enhance the flow later, the sonarqube stuff is a bit of a mess anyway.

bleggett replied:
And I would point out, there's nothing to "start over" with at this point, this is just collecting input, they can just modify the command and CLI flags they're using and rerun it in a matter of seconds if they actually gave us the wrong info.

bleggett commented on amazonlinux-provision.sh:
```diff
@@ -0,0 +1,279 @@
+#!/bin/bash
```
I actually need to remove this script, I created it some time ago but have not tested it, and it's OOS for this PR.

I shouldn't think we'd need different scripts for RPM-based distros.

bleggett commented:
>This is looking great. Please take a look at my comments.
Also I still see quite a few bash scripts. Do we want to tackle them as part of the next round of effort while this PR makes the journey to master?

There are a bunch of trivial ones specific to Chef and scenario1, these probably aren't worth reworking:
- terraform/provisioners/jenkinscli (used in scenario 1 only)

but only a few big/complex ones left in the main installer/terraform path:
- configureESEndpoint.sh
- modifyCodebase.sh
- scmpush.sh

rajeevr2715 commented on installer/configurators/jenkins.py:
```diff
@@ -0,0 +1,78 @@
+        " Please make sure that you have the ssh login user names of jenkins and bitbucket servers."
+    )
+    # Check if user dropped the jenkins key into the installer root (scenario 1 only)
+    if not Path.is_file(jenkins_pem):
```
Got error on this line, seems it should be "if not Path(jenkins_pem).is_file()"

bleggett replied:
Correct, fixed.

rajeevr2715 commented on installer/configurators/jenkins.py:
```diff
@@ -0,0 +1,78 @@
+    ssh_port = "22"
+
+    replace_tfvars('dockerizedJenkins', 'false', get_tfvars_file(), quoteVal=False)
+    update_jenkins_terraform(elb, userpw, ip, sshuser, ssh_port, secgrp, subnet)
```
7 parameters, but the actual function has only 6. Also, we might need jenkins_public_ip to ssh. Elb might have different IP each time.

bleggett replied:
So we don't even need an `elb` in this context, right? It's useless, we only want something we can connect to the Jenkins instance with.

We asked for both `elb` and public IP in the old installer, I confess I have no idea why we need both, just included both here (and was in the middle of removing one in favor of the other, which is why the args are jacked)

bleggett replied:
Confirmed, `jenkins_public_ip` was removed because it was unused, we only need to ask for an `endpoint` which can be the public DNS or public IP of the Jenkins instance, don't care which.

This will be stored in the Terraform variable `jenkins_elb` because that's the variable most scripts reference but it doesn't have to be an ELB.

bleggett replied:
Updated code to simply ask for a publicly accessible IP or a publicly resolvable DNS name, either should work.

rajeevr2715 replied:
Ok. If user provides public IP only, then ssh will work. But the jenkins endpoint will have to pointed to some port like 8080.

bleggett replied:
I'm not quite sure what you mean?

Yes, if something needs to hit a specific port, it will need to append that port to the IP.

rajeevr2715 replied:
For example - Jenkins endpoint: 54.XXX.XXX.XX7:8080
In the chefProvision, it will connect through Jenkins END point using port 22, So it will become, 
54.XXX.XXX.XX7:8080:22

connection {
    host = "${lookup(var.jenkinsservermap, "jenkins_elb")}"
    user = "${lookup(var.jenkinsservermap, "jenkins_ssh_login")}"
    port = "${lookup(var.jenkinsservermap, "jenkins_ssh_port")}"
    type = "ssh"
    private_key = "${file("${lookup(var.jenkinsservermap, "jenkins_ssh_key")}")}"
  }

bleggett replied:
Alright, that's scenario1 stuff, let me try and find all the spots in scenario1 where we're doing this and fix/verify.

bleggett replied:
Do we use a nonstandard Jenkins port in our internal use of scenario1?

If not, then I'm gonna hardcode it to the default like we do the SSH port in the existing code, since that could also change and we don't bother with making it configurable.

bleggett replied:
Alright, changed the installer to optionally accept a default port for Jenkins

`jenkins_rawendpoint` will be `<jenkins_endpoint>` and will be used for the SSH in the`chefProvisioner.tf` connection, since that's literally the only spot it's needed.

`jenkins_elb` will be `<jenkins_endpoint>:<jenkins_defaultport>` and will be used for everything else.

rajeevr2715 replied:
ok, then check_jenkins_user(endpoint, userpw) should be appended with defaultport

bleggett replied:
Good catch, done.

rajeevr2715 commented on installer/terraform/chefProvisioner.tf:
```diff
@@ -78,7 +96,7 @@ resource "null_resource" "configureExistingJenkinsServer" {
 
   connection {
-    host = "${lookup(var.jenkinsservermap, "jenkins_public_ip")}"
+    host = "${lookup(var.jenkinsservermap, "jenkins_elb")}"
```
we might need jenkins_public_ip to ssh. Elb might have different IP each time.

bleggett replied:
See prev comments.

rajeevr2715 commented on installer/configurators/jenkins.py:
```diff
@@ -0,0 +1,78 @@
+        )
+
+    # Make sure the PEM the user dropped in has the right perms
+    subprocess.call('sudo chmod 400 {0}'.format(jenkins_pem))
```
We might need shell=True.

bleggett replied:
I'm dropping the `sudo`, it's pointless.

If the user running this script isn't already the owner of the file, running `sudo chmod 400` will literally prevent them from using it and the installer will fail anyway because of bad Unix perms.

If the user running this script IS the owner of the file, they don't need the `sudo`.

rajeevr2715 replied:
ok. Getting this error, FileNotFoundError: [Errno 2] No such file or directory: 'chmod 400.....

Tried using subprocess.call('chmod 400 {0}'.format(jenkins_pem),shell=True), then the error gone.

bleggett replied:
Oh, duh I'm an idiot. `call` expects a string array where the first arg is the executable. Fixed.

rajeevr2715 commented on installer/configurators/bitbucket.py:
```diff
@@ -0,0 +1,68 @@
+    ]
+
+    try:
+        output = subprocess.check_output(cmd)
```
We might need "output = str(subprocess.check_output(cmd))", since it is checked against bytes.

bleggett replied:
Good call, added.

rajeevr2715 commented on installer/cli/scenario1cli.py:
```diff
@@ -0,0 +1,144 @@
+# TODO Why do we ask for all this Jenkins junk for scenario1? We really should just remove Jenkins from scenario 1.
+# If you are gonna BYOS you should install plugins and configure it yourself
+@click.option(
+    '--jenkins_userpass',
```
--jenkins_userpass xxxxx xxxxxx is working fine. But if it not used as a parameter, what should be the user input for jenkins username and password? I had given 

> Jenkins userpass [()]: xxxxx xxxxxx

But its failing.

bleggett replied:
With what error?


Might be a shell thing, might need to do `username\ pass` to escape the space character in the shell.

rajeevr2715 replied:
If I enter it like, 
Jenkins userpass [()]: username password
 print(jenkins_userpass) would be like this, 
('u', 's', 'e', 'r', 'n', 'a', 'm', 'e', ' ', 'p', 'a', 's', 's', 'w', 'o', 'r', 'd')

bleggett replied:
Added a workaround for this, it's a Click bug: https://github.com/pallets/click/issues/532

rajeevr2715 replied:
Ok, But the parameter way failed. --jenkins_userpass username password

After, 
print(userpw)

the result is,

[usernamepassword]

rajeevr2715 commented on installer/cli/scenario1cli.py:
```diff
@@ -0,0 +1,144 @@
+# Sonarqube (standalone)
+@click.option(
+    '--sonarqube/--no-sonarqube',
+    default=False
```
Seems by default its False. but it is still asking for --sonarqube_endpoint, --sonarqube_userpass etc.

bleggett replied:
Good catch, fixed in `click_required.py`

Fixed it for one clause and not the other.

rajeevr2715 commented on installer/terraform/scripts/injectingToJenkinsfile.sh:
```diff
@@ -6,15 +8,15 @@ scm_type=$2
+      file=$(find "$d" -type f -name "$file_name")
       if [[ "$file" =~ Jenkinsfile* ]] ; then
-        jenkinsfile_arr=(${jenkinsfile_arr[@]} "$file")
+        jenkinsfile_arr=("${jenkinsfile_arr[@]}" "$file")
```
Got one error, `^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): Executing: ["/bin/sh" "-c" "./scripts/injectingToJenkinsfile.sh benscen3-gitlab-alb-1355637415.us-east-1.elb.amazonaws.com gitlab"]
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): sed: can't read builds/delete-serverless-service-build-pack/Jenkinsfile
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): builds/delete-serverless-service-build-pack/Jenkinsfile_Cleanup_CF_Dist: No such file or directory
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): sed: can't read builds/delete-serverless-service-build-pack/Jenkinsfile
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): builds/delete-serverless-service-build-pack/Jenkinsfile_Cleanup_CF_Dist: No such file or directory
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): sed: can't read builds/delete-serverless-service-build-pack/Jenkinsfile
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): builds/delete-serverless-service-build-pack/Jenkinsfile_Cleanup_CF_Dist: No such file or directory
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): sed: can't read builds/delete-serverless-service-build-pack/Jenkinsfile
^[[0m^[[0mnull_resource.postJenkinsConfiguration (local-exec): builds/delete-serverless-service-build-pack/Jenkinsfile_Cleanup_CF_Dist: No such file or directory
^[[0m^[[1mnull_resource.postJenkinsConfiguration: Creation complete after 32s (ID: 536112423785532212)^[[0m^[[0m
`
Seems we need to revert it to, jenkinsfile_arr=(${jenkinsfile_arr[@]} "$file")

Please verify

bleggett replied:
I can't really test or diagnose this locally or on my end, I will remove the quotes, let me know if it fixes this or if there's a different error.

rajeevr2715 replied:
ok. I had tested in scenario3 and got this issue.  All the jenkins job failed. Let me re test.

bleggett replied:
Yep you're right, I saw `depends_on = ["null_resource.configureJenkins"]` and misremembered that as being a scenario1-only step.

I'm starting to wonder if it might not be better to create a separate set of Terraform scripts for Scenario1 so we don't need all these conditionals and branching logic just to support a use case most OSS users won't use.

Might be cleaner, not sure.

rajeevr2715 commented:
@bleggett Seems like the terraform apply is not running as background job. When an interruption happends in command line, the script execution stops. There was no logs written to the log file after the interruption. Might be need to run the function in background.

bleggett commented:
@rajeevr2715 Yeah I removed the backgrounding, I don't see the point in doing it, it complicates things and accomplishes very little besides hiding Terraform output and making you have to go check a logfile to see failures rather than just see your output.

There are no logs written after Terraform fails/is interrupted because there's no point in doing so after that point, if Terraform fails the installer should fail immediately as well.

rajeevr2715 commented:
@bleggett So we are not going to display final stack details like ,

echo " Please use the following values for checking out Jazz" by calling this script "python ./scripts/generatestackdetails.py" ?

rajeevr2715 commented on installer/terraform/chefProvisioner.tf:
```diff
@@ -136,7 +155,7 @@ resource "null_resource" "configureJenkins" {
   #Jenkins Cli process
   provisioner "local-exec" {
-    command = "bash ${var.configureJenkinsCE_cmd} ${var.dockerizedJenkins == 1 ? join(" ", aws_lb.alb_ecs_jenkins.*.dns_name) : lookup(var.jenkinsservermap, "jenkins_elb") } ${var.cognito_pool_username} ${var.dockerizedJenkins} ${var.scmgitlab == 1 ? join(" ", aws_lb.alb_ecs_gitlab.*.dns_name) : lookup(var.scmmap, "scm_elb") } ${lookup(var.scmmap, "scm_username")} ${lookup(var.scmmap, "scm_passwd ...
+    command = "bash ${var.configureJenkinsCE_cmd} ${var.dockerizedJenkins == 1 ? join(" ", aws_lb.alb_ecs_jenkins.*.dns_name) : lookup(var.jenkinsservermap, "jenkins_elb") } ${var.cognito_pool_username} ${var.dockerizedJenkins} ${var.scmgitlab == 1 ? join(" ", aws_lb.alb_ecs_gitlab.*.dns_name) : lookup(var.scmmap, "scm_elb") } ${lookup(var.scmmap, "scm_username")} ${lookup(var.scmmap, "scm_passwd ...
```
Missed ${var.acl_db_username} ${var.acl_db_password}

bleggett replied:
Yep, added.

rajeevr2715 commented on installer/terraform/chefProvisioner.tf:
```diff
@@ -112,14 +130,15 @@ resource "null_resource" "configureExistingJenkinsServer" {
   query = {
     passwd = "${var.cognito_pool_password}"
-    ip = "${aws_lb.alb_ecs_gitlab.dns_name}"
-    gitlab_admin = "${lookup(var.scmmap, "scm_username")}"
```
By default the gitlab username is root, so we have to update the username https://github.com/tmobile/jazz-installer/blob/9e7e14f0467e40537ef8f9892c7439bec349f7a7/installscripts/jazz-terraform-unix-noinstances/scripts/configure-gitlab.sh#L9

Also, the username should be  replaced with '[^a-zA-Z0-9_-]', '-' 
- https://github.com/tmobile/jazz-installer/blob/9e7e14f0467e40537ef8f9892c7439bec349f7a7/installscripts/wizard/scenarios/support/jazz_gitlab.py#L26

bleggett replied:
What's the point of changing the name of the default `root` user? I'm assuming "security", but that's unlikely to make a real difference to anyone running a dictionary attack or whatever against the login form, as long as we use a nondefault password.

rajeevr2715 replied:
@bleggett Earlier what we did is, update the root user to a specific username and used that for login. Now Gitlab login is not working using the email and password. By default there wont be any email in gitlab and username would be root.  In the final stack output it is the email that is displaying to login. Also in the jenkins credentials, username is email. So all the builds will fail.

bleggett replied:
I thought Jenkins was using a personal access token, not username/password?

bleggett replied:
We should revisit this, because it feels like the way it's currently being done is wrong, but for the time being I have made it work like it does currently.

bleggett commented:
> @bleggett So we are not going to display final stack details like ,
> 
> echo " Please use the following values for checking out Jazz" by calling this script "python ./scripts/generatestackdetails.py" ?

Nope, not if the process fails, what's the point then?

You're right that I forgot to integrate `generatestackdetails.py` on a successful run though, lemme do that real quick.

rajeevr2715 commented:
> > @bleggett So we are not going to display final stack details like ,
> > echo " Please use the following values for checking out Jazz" by calling this script "python ./scripts/generatestackdetails.py" ?
> 
> Nope, not if the process fails, what's the point then?
> 
> Note that most users simply don't need `stack_details.json`, every bit of information in it can be seen by running `terraform output`, we probably should move away from using it, I don't see that it adds significant value beyond `terraform output`.
> 
> What's more, `terraform output -json` will give you a json of all the Terraform output values, which is currently everything that's in `stack_details.json`.
> 
> Do we have some internal process that relies on `stack_details.json` being in a specific format? If so, let me know.
> 
> Otherwise I'm simply going to dump the result of `terraform output -json` to a file named `stack_details.json`, there's no reason to complicate it beyond that.

yeah, we need this stack details as json to invoke internal uiautomation related internal jobs.

{
    "Jenkins ELB": "http://XXX",
    "Jenkins Username": "XXX",
    "Jenkins Password": "XXXX",
    "Jazz Home": "http://XXX",
    "Jazz Admin Username": "XXX",
    "Jazz Admin Password": "XXX",
    "Region": "us-east-1",
    "Jazz API Endpoint": "https://lXXXX",
    "Gitlab Home": "http://sXXXX",
    "Gitlab Username": "sXXXX",
    "Gitlab Password": "XXXX",
   "Sonar Home": "XXXX",
    "Sonar Username",
    "Sonar Password",
     "Bitbucket ELB",
     "Bitbucket Username",
      "Bitbucket Password"
}

https://github.com/tmobile/jazz-installer/blob/9e7e14f0467e40537ef8f9892c7439bec349f7a7/installscripts/jazz-terraform-unix-noinstances/scripts/generatestackdetails.py

rajeevr2715 commented on installer/terraform/scmProvisioner.tf:
```diff
@@ -6,22 +6,27 @@ resource "null_resource" "createProjectsInBB" {
   depends_on = ["null_resource.copyJazzBuildModule"]
   provisioner "local-exec" {
-    command = "${var.pushInstallervars_cmd} ${lookup(var.scmmap, "scm_username")} ${urlencode(lookup(var.scmmap, "scm_passwd"))} ${var.scmgitlab == 1 ? join(" ", aws_lb.alb_ecs_gitlab.*.dns_name) : lookup(var.scmmap, "scm_elb") } ${lookup(var.scmmap, "scm_pathext")} ${var.cognito_pool_username}"
+    command = "${var.pushInstallervars_cmd} ${lookup(var.scmmap, "scm_username")} ${urlencode(lookup(var.scmmap, "scm_passwd"))} ${var.scmgitlab == 1 ? join(" ", aws_lb.alb_ecs_gitlab.*.dns_name) : lookup(var.scmmap, "scm_elb") } ${lookup(var.scmmap, "scm_pathext")} ${var.cognito_pool_username} ${lookup(var.scmmap, "scm_type")} ${var.scmgitlab == 1 ? join(" ", data.external.gitlabconfig.*.result. ...
```
@bleggett Got this error ^[[0m^[[0mnull_resource.configureJazzBuildModule (local-exec): Cloning into 'jazz-build-module'...
^[[0m^[[0mnull_resource.configureJazzBuildModule (local-exec): cp: cannot stat ‘/installer/terraform/provisioners/cookbooks/jenkins/files/default/jazz-installer-vars.json’: No such file or directory
^[[0m^[[0mnull_resource.configureJazzBuildModule (local-exec): fatal: pathspec 'jazz-installer-vars.json' did not match any files

bleggett replied:
Does this fail the run? If it doesn't I might have missed it in the output of my scenario3 testing.

bleggett replied:
Ah fixed. Somehow the script was using an env var to root the path, when it didn't really need to at all, can use a relative path like the other scripts do.

bleggett commented:
> > > @bleggett So we are not going to display final stack details like ,
> > > echo " Please use the following values for checking out Jazz" by calling this script "python ./scripts/generatestackdetails.py" ?
> > 
> > 
> > Nope, not if the process fails, what's the point then?
> > Note that most users simply don't need `stack_details.json`, every bit of information in it can be seen by running `terraform output`, we probably should move away from using it, I don't see that it adds significant value beyond `terraform output`.
> > What's more, `terraform output -json` will give you a json of all the Terraform output values, which is currently everything that's in `stack_details.json`.
> > Do we have some internal process that relies on `stack_details.json` being in a specific format? If so, let me know.
> > Otherwise I'm simply going to dump the result of `terraform output -json` to a file named `stack_details.json`, there's no reason to complicate it beyond that.
> 
> yeah, we need this stack details as json to invoke internal uiautomation related internal jobs.
> 
> {
> "Jenkins ELB": "http://XXX",
> "Jenkins Username": "XXX",
> "Jenkins Password": "XXXX",
> "Jazz Home": "http://XXX",
> "Jazz Admin Username": "XXX",
> "Jazz Admin Password": "XXX",
> "Region": "us-east-1",
> "Jazz API Endpoint": "https://lXXXX",
> "Gitlab Home": "http://sXXXX",
> "Gitlab Username": "sXXXX",
> "Gitlab Password": "XXXX",
> "Sonar Home": "XXXX",
> "Sonar Username",
> "Sonar Password",
> "Bitbucket ELB",
> "Bitbucket Username",
> "Bitbucket Password"
> }
> 
> https://github.com/tmobile/jazz-installer/blob/9e7e14f0467e40537ef8f9892c7439bec349f7a7/installscripts/jazz-terraform-unix-noinstances/scripts/generatestackdetails.py

Alright that's what I was afraid of, will add generatestackdetails back in.

rajeevr2715 commented:
@bleggett I have tried scenario3 and got an error on jenkins builds.
Error details - 
> Cloning the remote Git repository
> Cloning repository http://**REPLACEME**/scm/slf/jazz_acl.git

Installer vars configuration value for Repository was not replaced.

> "REPOSITORY": {
>         "BASE_URL" : "REPLACEME",

Also SCM type and CAS was not changed, 

> SCM": {
>         "TYPE": "bitbucket",
>         "PRIVATE_TOKEN" : "XXXXXXX",
>         "CAS_NAMESPACE_ID": ""
>     },

rajeevr2715 commented:
@bleggett Tried scenario1 and found, the below json structure also not got replaced.

> "API": {
>             "PROD": {
>                 "*": "{AWS_PROD_API_ID_DEFAULT}",
>                 "jazz_*": "{AWS_PROD_API_ID_JAZZ}"
>             },
>             "STG": {
>                 "*": "{AWS_STG_API_ID_DEFAULT}",
>                 "jazz_*": "{AWS_STG_API_ID_JAZZ}"
>             },
>             "DEV": {
>                 "*": "{AWS_DEV_API_ID_DEFAULT}"
>             }

bleggett commented:
> @bleggett I have tried scenario3 and got an error on jenkins builds.
> Error details -
> 
> > Cloning the remote Git repository
> > Cloning repository http://**REPLACEME**/scm/slf/jazz_acl.git
> 
> Installer vars configuration value for Repository was not replaced.
> 
> > "REPOSITORY": {
> > "BASE_URL" : "REPLACEME",
> 
> Also SCM type and CAS was not changed,
> 
> > SCM": {
> > "TYPE": "bitbucket",
> > "PRIVATE_TOKEN" : "XXXXXXX",
> > "CAS_NAMESPACE_ID": ""
> > },

Yup, this is related to features that were merged after I tested this branch, will work them in and retest.

bleggett commented:
> @bleggett I have tried scenario3 and got an error on jenkins builds.
> Error details -
> 
> > Cloning the remote Git repository
> > Cloning repository http://**REPLACEME**/scm/slf/jazz_acl.git
> 
> Installer vars configuration value for Repository was not replaced.
> 
> > "REPOSITORY": {
> > "BASE_URL" : "REPLACEME",
> 
> Also SCM type and CAS was not changed,
> 
> > SCM": {
> > "TYPE": "bitbucket",
> > "PRIVATE_TOKEN" : "XXXXXXX",
> > "CAS_NAMESPACE_ID": ""
> > },

This looks to be related to new code that has been merged into upstream since I last tested scenario3, will reintegrate.

bleggett commented:
>  "CAS_NAMESPACE_ID": ""

Hey so I removed CAS_NAMESPACE_ID because I could find no references to that variable in `jazz`, `jazz-content`, `jazz-internal` and all `jazz-installer` does is set it.

Where is this variable actually used? It doesn't appear as if it is, which I think is why I originally removed it.

bleggett commented:
> @bleggett Tried scenario1 and found, the below json structure also not got replaced.
> 
> > "API": {
> > "PROD": {
> > "_": "{AWS_PROD_API_ID_DEFAULT}",
> > "jazz__": "{AWS_PROD_API_ID_JAZZ}"
> > },
> > "STG": {
> > "_": "{AWS_STG_API_ID_DEFAULT}",
> > "jazz__": "{AWS_STG_API_ID_JAZZ}"
> > },
> > "DEV": {
> > "*": "{AWS_DEV_API_ID_DEFAULT}"
> > }

Yup dropped the standalone shell script and moved those updates into `jenkins.tf` to be updated by `modifyPropertyFile.sh`, will see why they're not being updated.

bleggett commented:
> > @bleggett Tried scenario1 and found, the below json structure also not got replaced.
> > > "API": {
> > > "PROD": {
> > > "_": "{AWS_PROD_API_ID_DEFAULT}",
> > > "jazz__": "{AWS_PROD_API_ID_JAZZ}"
> > > },
> > > "STG": {
> > > "_": "{AWS_STG_API_ID_DEFAULT}",
> > > "jazz__": "{AWS_STG_API_ID_JAZZ}"
> > > },
> > > "DEV": {
> > > "*": "{AWS_DEV_API_ID_DEFAULT}"
> > > }
> 
> Yup dropped the standalone shell script and moved those updates into `jenkins.tf` to be updated by `modifyPropertyFile.sh`, will see why they're not being updated.

Ah, they were never updated because `modifyPropertyFile` expects to be passed a key to replace the value on, and does not expect to be passed a stub value to replace, which is what the old script was doing.

It's unfortunate we chose update some properties in `jazz-installer.vars.json`by key and some by value, as that creates unnecessary complexity like this and as far as I can tell is purely cosmetic.

bleggett commented:
Updated with latest new feature merge `develop`, hopefully that didn't break anything.

bleggett commented:
1. Corrected everything brought up thus far (except for "CAS_NAMESPACE_ID", since as mentioned that seems to be unused), scenario3 seems to be working.

2. However after merging latest develop into this branch, all scenarios (in this branch as well as develop) fail for me with this error: 


```
aws_rds_cluster.casbin: error creating RDS cluster: InvalidVPCNetworkStateFault: Availability zone '[us-east-1c]' is unavailable in this region, please choose another zone set.
```

According to `aws ec2 describe-availability-zones --region us-east-1` `us-east-1c` _IS_ definitely available to me, any ideas as to what might be causing this?

raghits commented:
> @rajeevr2715
> 
> 1. Corrected everything brought up thus far (except for "CAS_NAMESPACE_ID", since as mentioned that seems to be unused based on grepping all our repos, so I'm excising it unless someone can show where it's being used), scenario3 seems to be working.
> 2. However after merging latest develop into this branch, all scenarios (in this branch as well as develop) fail for me with this error:
> 
> ```
> aws_rds_cluster.casbin: error creating RDS cluster: InvalidVPCNetworkStateFault: Availability zone '[us-east-1c]' is unavailable in this region, please choose another zone set.
> ```
> According to `aws ec2 describe-availability-zones --region us-east-1`, `us-east-1c` _IS_ definitely available to me, any ideas as to what might be causing this?

`aws rds describe-source-regions` doesn't show us-east-1 or us-east-2? Does that explain it? We are talking RDS here not EC2.

rajeevr2715 commented:
Created one PR to handle the issue in RDS - https://github.com/tmobile/jazz-installer/pull/525
Availability zones specified in RDS resources should be present in the subnets which are specified in subnet group in RDS.

rajeevr2715 commented:
@bleggett tested scenario1 and found 

> SCM": {
       "TYPE": "{SCM_TYPE}",
It should be updated to "bitbucket" or by default it should be "bitbucket" .

bleggett commented:
> @bleggett tested scenario1 and found
> 
> > SCM": {
> > "TYPE": "{SCM_TYPE}",
> > It should be updated to "bitbucket" or by default it should be "bitbucket" .

You're right, I switched the method to by-placeholder-value and the bitbucket logic doesn't update. Fixed.

rajeevr2715 commented on installer/terraform/scripts/configure-gitlab.py:
```diff
@@ -0,0 +1,30 @@
+
+    gl = gitlab.Gitlab('http://{0}'.format(query['gitlab_ip']), api_version=4, private_token=token)
+
+    group = gl.groups.create({'name': 'SLF', 'path': 'slf', 'description': 'Jazz framework, templates and services'})
```
@bleggett Need this repo to be created along with SLF.

>  '{"name":"CAS","path":"cas", "description": "User created services repository"}')

bleggett replied:
Oh so we need the group but not the variable, ok.

rajeevr2715 commented:
@bleggett @devsatishm I had tested the code change in dockerized and non dockerized scenarios and its working fine.
